core/templates/services/playthrough.service.ts

```diff
@@ -1,334 +1,334 @@
+   *
+   * Finally, the path of visited states is reset to a value of [ N ], in hopes
+   * that the exact same cycle is discovered enough times to be considered an
+   * issue.
+   */
+  recordStateTransition(destStateName: string): void {
+    if (this.currStateName() === destStateName) {
+      return;
+    }
+    if (this.pathOfVisitedStates.includes(destStateName)) {
+      const cycleOfVisitedStates = (
+        this.makeCycle(this.pathOfVisitedStates.indexOf(destStateName)));
+      if (angular.equals(this.cycleOfVisitedStates, cycleOfVisitedStates)) {
+        this.numLoops += 1;
+      } else {
+        this.cycleOfVisitedStates = cycleOfVisitedStates;
+        this.numLoops = 1;
+      }
+      this.pathOfVisitedStates.length = 0;
+    }
+    this.pathOfVisitedStates.push(destStateName);
+  }
+
+  generateIssueCustomizationArgs(): CyclicStateTransitionsCustomizationArgs {
+    return {
+      state_names: {value: this.cycleOfVisitedStates}
+    };
+  }
+
+  private makeCycle(collisionIndex: number): string[] {
+    const collision = this.pathOfVisitedStates[collisionIndex];
+    const cycleWithNoCollision = this.pathOfVisitedStates.slice(collisionIndex);
+    return [...cycleWithNoCollision, collision];
+  }
+
+  private currStateName(): string {
+    return this.pathOfVisitedStates[this.pathOfVisitedStates.length - 1];
+  }
+}
+
+class EarlyQuitTracker {
+  private stateName: string = " ";
+  private expDurationInSecs: number | null = null;
+
+  foundAnIssue(): boolean {
+    return (
+      this.expDurationInSecs !== null &&
+      this.expDurationInSecs < ServicesConstants.EARLY_QUIT_THRESHOLD_IN_SECS);
+  }
+
+  recordExplorationQuit(stateName:string , expDurationInSecs: number | null): void {
+    this.stateName = stateName;
+    this.expDurationInSecs = expDurationInSecs;
+  }
+
+  generateIssueCustomizationArgs(): EarlyQuitCustomizationArgs {
+    return {
+      state_name!: {value: this.stateName},
+      time_spent_in_exp_in_msecs: {value: this.expDurationInSecs! * 1000},
+    };
+  }
+}
+
+class MultipleIncorrectAnswersTracker {
+  private currStateName: string;
+  private numTries: number;
+
+  constructor(initStateName: string) {
+    this.currStateName = initStateName;
+    this.numTries = 0;
+  }
+
+  foundAnIssue(): boolean {
+    return this.numTries >= ServicesConstants.NUM_INCORRECT_ANSWERS_THRESHOLD;
+  }
+
+  recordStateTransition(destStateName: string): void {
+    if (this.currStateName === destStateName) {
+      this.numTries += 1;
+    } else {
+      this.currStateName = destStateName;
+      this.numTries = 0;
+    }
+  }
+
+  generateIssueCustomizationArgs(
+  ): MultipleIncorrectSubmissionsCustomizationArgs {
+    return {
+      state_name: {value: this.currStateName},
+      num_times_answered_incorrectly: {value: this.numTries},
+    };
+  }
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class PlaythroughService {
+  private explorationId: string = " ";

```

Comment:
An empty string is not a good replacement for `null`. You need to either get rid of the `null` altogether or allow it explicitely.

Reply:
Done

Reply:
Again empty string is not a good replacement for `null`.

Reply:
It is showing "null is not assignable to type string".